core/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/QualificationAppInfo.scala

```diff
@@ -241,13 +242,49 @@ class QualificationAppInfo(
     stages.map { stageId =>
       val stageTaskTime = stageIdToTaskEndSum.get(stageId)
         .map(_.totalTaskDuration).getOrElse(0L)
+      val transitionsTime = stageIdToGpuCpuTransitions.getOrElse(stageId, 0) match {
+        case gpuCpuTransitions if gpuCpuTransitions > 0 =>
+          // Duration to transfer data from GPU to CPU and vice versa.
+          // Assuming it's a PCI-E Gen3, but also assuming that some of the result could be
+          // spilled to disk.
+          // Duration in Spark metrics is in millisecond, so multiply this by 1000 to make
+          // it consistent
+          val totalBytesRead =
+            stageIdToTaskEndSum.get(stageId).map(_.totalbytesRead).getOrElse(0L).toDouble
+          if (totalBytesRead > 0) {
+            val fallback_duration = (totalBytesRead / QualificationAppInfo.CPU_GPU_TRANSFER_RATE) *
+              QualificationAppInfo.SECONDS_TO_MILLISECONDS * gpuCpuTransitions
+            fallback_duration.toLong
+          } else {
+            0L
+          }
+        case _ => 0L
+      }
+      // Update totaltaskduration of stageIdToTaskEndSum to include transitions time
+      val stageIdToTasksMetrics = stageIdToTaskEndSum.get(stageId).orElse(None)
+      if (stageIdToTasksMetrics.isDefined) {
+        stageIdToTasksMetrics.get.totalTaskDuration += transitionsTime
+      }
       StageQualSummaryInfo(stageId, allSpeedupFactorAvg, stageTaskTime,
-        eachStageUnsupported, estimated)
+        eachStageUnsupported + transitionsTime, estimated)
     }.toSet
   }
 
   def summarizeStageLevel(execInfos: Seq[ExecInfo], sqlID: Long): Set[StageQualSummaryInfo] = {
     val (allStagesToExecs, execsNoStage) = getStageToExec(execInfos)
+
+    // Get the total number of transitions between CPU and GPU for each stage and
+    // store it in a Map.
+    allStagesToExecs.foreach { case (stageId, execs) =>
```

Comment:
this isn't looking at execsNoStages. I can't remember when that all happens but could miss transitions

Reply:
We skip execsNoStages as there is no way we could map Exec to any stageId. We will not know which stage to include if there are any transitions. I had created an issue to improve qual tool earlier. Basically we miss some Execs in stages where the Execs don't have any metrics and we map stages to Execs based on accumulator Id.
issue - https://github.com/NVIDIA/spark-rapids-tools/issues/472
```
Some Execs do not have metrics populated, so while calculating the number of Execs within a stage we miss these. We map stageToExec based on accumulatorId and since metrics are not populated, we miss adding those Execs to that corresponding stage.
```